feat: ooo exec

makramkd opened the pull request:
## Motivation

We want to support executing CCIP messages out of order.

## Solution

* Add an additional extraArgs field, called `allowOutOfOrderExecution`, that can be specified by the caller.
* Add an additional configuration field to the EVM2EVMOnRamp, `enforceOutOfOrder`, which can be set to `true` in order to enforce that `allowOutOfOrderExecution` be set to `true`.

## Notes

I recommend viewing the diff with "Hide whitespace": 
<img width="233" alt="Screenshot 2024-04-17 at 3 46 57 PM" src="https://github.com/smartcontractkit/ccip/assets/6530589/70133f2e-e70f-4f41-98df-76292ca8cb0f">

makramkd committed:
feat: ooo exec

makramkd committed:
add e2e test

makramkd committed:
update integration tests

makramkd committed:
interleave sequenced and nonsequenced

makramkd commented on core/services/ocr2/plugins/ccip/ccipexec/ocr2.go:
```diff
@@ -331,8 +331,8 @@ func (r *ExecutionReportingPlugin) buildBatch(
-		// Check expected nonce is valid
-		if msg.Nonce != expectedNonces[msg.Sender] {
+		// Check expected nonce is valid for sequenced messages.
+		if msg.Strict && msg.Nonce != expectedNonces[msg.Sender] {
```
This is hacky and won't work on pre-1.5.0 deployments since `strict` is always set to `false` there.

Its possible that this batch building logic may need to live behind an interface instead.

RensR commented:
This needs to go into 1.2 lane contracts, not 1.5 right? Or are we waiting for 1.5 for this feature?

matYang commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol:
```diff
@@ -359,14 +361,24 @@ contract EVM2EVMOnRamp is IEVM2AnyOnRamp, ILinkAvailable, AggregateRateLimiter,
+    }
+
+    bytes4 extraArgsTag = bytes4(extraArgs);
+    if (extraArgsTag == Client.EVM_EXTRA_ARGS_V1_TAG) {
```
nit: default cause can be V2

makramkd replied:
By default clause you mean first clause in the if statement?

RensR replied:
Yeah if we expect people to use it, which we do, it should be the first

makramkd replied:
Made it the first clause.

matYang commented on contracts/src/v0.8/ccip/libraries/Client.sol:
```diff
@@ -36,4 +36,18 @@ library Client {
     return abi.encodeWithSelector(EVM_EXTRA_ARGS_V1_TAG, extraArgs);
   }
+
+  // bytes4(keccak256("CCIP EVMExtraArgsV2"));
```
Since `Client.sol` is customer-facing, let's add a comment that clearly states ArgsV1 is being deprecated. I guess it's gonna be a bit complex when some lanes support ArgsV2, and some don't. For now customers will need to read the docs to figure out which which lane supports which.

When every lane upgrades, we should sunset _argsToBytes for v1.

```
error DeprecatedExtraArgsVersion();

function _argsToBytes(EVMExtraArgsV1 memory extraArgs) internal pure returns (bytes memory bts) {
    revert DeprecatedExtraArgsVersion();
}
```

makramkd replied:
Just to confirm, you want `_argsToBytes` for V1 args to revert as you mentioned above in this PR? Or would this be in a future `@chainlink/contracts` that would just change `Client.sol`?

RensR replied:
We're not deprecating v1, just adding v2 right? We cannot break current customers

RensR replied:
> when some lanes support ArgsV2, and some don't

This will not happen. It will roll out to all lanes at the same time

makramkd replied:
> We cannot break current customers

That is correct. We won't do that with this upgrade.

matYang commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMOffRamp.sol:
```diff
@@ -273,76 +273,93 @@ contract EVM2EVMOffRamp is IAny2EVMOffRamp, AggregateRateLimiter, ITypeAndVersio
-          // it tells us there are still messages inflight for v1 offramp
-          emit SkippedSenderWithPreviousRampMessageInflight(message.nonce, message.sender);
-          continue;
+      Internal.MessageExecutionState newState;
```
let's extract this nonce logic into own function

makramkd replied:
Already extracted the common logic between both code paths into `_doExecute`. I feel like its simpler to keep the nonce logic inline due to the `continue`s but I don't feel too strongly about it.

makramkd replied:
Updated to a hopefully much simpler diff.

matYang commented on contracts/src/v0.8/ccip/libraries/Client.sol:
```diff
@@ -36,4 +36,18 @@ library Client {
+  bytes4 public constant EVM_EXTRA_ARGS_V2_TAG = 0x181dcf10;
+
+  /// @param gasLimit: gas limit for the callback on the destination chain.
+  /// @param sequenced: if true, the message can only be executed if the previous sequenced one was executed (whether successfully or not).
```
I like the word `sequenced`, but let's conf from GTM/PRD on the term.

We should also explain that `sequenced` means messages will be *attempted* in order, with no guarantee on msg success being in order. While `non-sequenced` means message may be `attempted` in random order, and this explicitly protects customer dapps from being permanently bricked due to problems like zk overflow

makramkd replied:
> this explicitly protects customer dapps from being permanently bricked due to problems like zk overflow

Does this need to be explicitly mentioned?

matYang commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol:
```diff
@@ -302,9 +302,11 @@ contract EVM2EVMOnRamp is IEVM2AnyOnRamp, ILinkAvailable, AggregateRateLimiter,
-      strict: false,
-      nonce: ++s_senderNonce[originalSender],
+      gasLimit: extraArgsV2.gasLimit,
+      strict: extraArgsV2.sequenced,
```
can we check nonce==0 instead of overloading this field? If not, pls add comments here and in Internal.EVM2EVMMessage that this field is being used for new purpose.

makramkd replied:
Fair enough, can do that.

makramkd committed:
CR comments

makramkd committed:
Merge branch 'ccip-develop' into feat/ccip/ooo

makramkd committed:
rename to allowOutOfOrder

makramkd committed:
bump wrappers

makramkd committed:
bump integration test

makramkd committed:
enforce flag using config

code size exceeds limit

RensR commented on contracts/src/v0.8/ccip/libraries/Client.sol:
```diff
@@ -36,4 +36,20 @@ library Client {
+  /// is not set, the message request will revert.
+  struct EVMExtraArgsV2 {
+    uint256 gasLimit;
+    bool allowOutOfOrder;
```
Can you make the name a bit more descriptive, like `allowOutOfOrderExecution`

makramkd replied:
Got the `allowOutOfOrder` from Pieter, IMO it's already quite verbose. I feel `orderedExecution` is quite descriptive and compact but I'm indifferent, let me ask if Pieter's okay with your suggestion.

makramkd replied:
Renamed to allowOutOfOrderExecution

RensR commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMOffRamp.sol:
```diff
@@ -288,76 +288,93 @@ contract EVM2EVMOffRamp is IAny2EVMOffRamp, AggregateRateLimiter, ITypeAndVersio
+          }
         }
+
+        (newState, returnData) = _doExecute(message, report.offchainTokenData[i], manualExecution, originalState);
```
The end of the if and the else are the same code. let's dedupe and remove the else

RensR commented on contracts/src/v0.8/ccip/test/e2e/End2End.t.sol:
```diff
@@ -17,6 +17,80 @@ contract E2E is EVM2EVMOnRampSetup, CommitStoreSetup, EVM2EVMOffRampSetup {
     deployOffRamp(s_commitStore, s_destRouter, address(0));
   }
 
+  function test_E2E_OrderedAndUnorderedInterleaved_Success() public {
```
imo this doesn't add much from an E2E perspective and I'd prefer the ordering logic to be tested in the offRamp tests over the E2E. This test will *always* compile on any contract change, so not adding too much here will reduce the compile/test times for whatever anyone is working on. Nothing here cannot be a normal unit test

RensR commented on contracts/src/v0.8/ccip/test/offRamp/EVM2EVMOffRamp.t.sol:
```diff
@@ -233,9 +233,10 @@ contract EVM2EVMOffRamp_execute is EVM2EVMOffRampSetup {
       )
     );
     // Nonce should increment on non-strict
-    assertEq(uint64(0), s_offRamp.getSenderNonce(address(OWNER)));
```
Missing unit tests for the new logic. I'd like a unit test where the same sender has 
unorder -> ordered -> unorderd -> ordered msgs being executed, where the order I gave here is the seq num ordering

makramkd replied:
Was testing it in `test_E2E_OrderedAndUnorderedInterleaved_Success` but will move to do unit tests in Offramp.t.sol instead 👍

makramkd replied:
Modified most tests to be fuzz tests based off of the `allowOutOfOrderExecution` logic (nonce = 0) in order to get as much coverage as possible. Also added a fuzz test that fuzzes ordering and asserts correct outcomes.

makramkd committed:
more tests

makramkd committed:
rename allowOutOfOrder to allowOutOfOrderExecution

makramkd committed:
simplify

makramkd committed:
Merge branch 'feat/ccip/ooo' into feat/ccip/ooo-enforce-via-config

makramkd committed:
Merge branch 'ccip-develop' into feat/ccip/ooo

makramkd committed:
Merge branch 'feat/ccip/ooo' into feat/ccip/ooo-enforce-via-config

makramkd committed:
simplify to a single flag

makramkd committed:
add test for no extraArgs provided

makramkd committed:
Merge branch 'ccip-develop' into feat/ccip/ooo

makramkd committed:
forge fmt

makramkd commented on contracts/scripts/native_solc_compile_all_ccip:
```diff
@@ -8,7 +8,7 @@ echo " └───────────────────────
 SOLC_VERSION="0.8.19"
 OPTIMIZE_RUNS=26000
-OPTIMIZE_RUNS_LOW=4000
+OPTIMIZE_RUNS_LOW=3650
```
Had to drop this to meet the size limit but can bump again if other changes that reduce contract size are merged before.

RensR commented on contracts/src/v0.8/ccip/test/offRamp/EVM2EVMOffRamp.t.sol:
```diff
@@ -358,10 +402,105 @@ contract EVM2EVMOffRamp_execute is EVM2EVMOffRampSetup {
+  function test_FuzzInterleavingOrderedAndUnorderedMessages_Success(
+    bool msg1Ordered,
+    bool msg2Ordered,
+    bool msg3Ordered
```
Prefer bool[3] and loops over hard coding all cases

makramkd replied:
Array-ified the test and made it `bool[7]` to get more coverage

RensR commented on contracts/src/v0.8/ccip/test/offRamp/EVM2EVMOffRamp.t.sol:
```diff
@@ -509,9 +665,12 @@ contract EVM2EVMOffRamp_execute is EVM2EVMOffRampSetup {
   }
 
-  function test_RouterYULCall_Revert() public {
+  function test_FuzzRouterYULCall_Revert(bool ordered) public {
```
Nit: please add an underscore after Fuzz in each test name

makramkd replied:
Good call, fixed and updated snaps.

makramkd committed:
rename fuzz and update snap

makramkd committed:
simplify fuzz test

makramkd committed:
add changeset

makramkd committed:
forge fmt

RensR commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol:
```diff
@@ -370,14 +374,34 @@ contract EVM2EVMOnRamp is IEVM2AnyOnRamp, ILinkAvailable, AggregateRateLimiter,
+        revert ExtraArgOutOfOrderExecutionMustBeTrue();
+      }
+      return extraArgsV2;
+    } else {
```
nit can remove the else statement

RensR commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol:
```diff
@@ -370,14 +374,34 @@ contract EVM2EVMOnRamp is IEVM2AnyOnRamp, ILinkAvailable, AggregateRateLimiter,
+    } else if (extraArgsTag == Client.EVM_EXTRA_ARGS_V2_TAG) {
+      Client.EVMExtraArgsV2 memory extraArgsV2 = abi.decode(extraArgs[4:], (Client.EVMExtraArgsV2));
+
+      if (s_dynamicConfig.enforceOutOfOrder && !extraArgsV2.allowOutOfOrderExecution) {
```
Let's swap the ordering of this check to prevent the storage read in the case allowOutOfOrderExecution is false

RensR commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol:
```diff
@@ -370,14 +374,34 @@ contract EVM2EVMOnRamp is IEVM2AnyOnRamp, ILinkAvailable, AggregateRateLimiter,
+  }
+
+  function _maybeEnforceOutOfOrderExecution(uint256 gasLimit) internal view returns (Client.EVMExtraArgsV2 memory) {
+    if (s_dynamicConfig.enforceOutOfOrder) {
```
can be replaced by 

return Client.EVMExtraArgsV2({gasLimit: gasLimit, allowOutOfOrderExecution: s_dynamicConfig.enforceOutOfOrder});

RensR replied:
Probably fine to inline at that point

RensR replied:
Wait, this isnt the behavior we want, we should revert if not set but required. This would allow changing the default behaviour of ccip through setting OOO on v1 args/no args

makramkd replied:
Ah yeah I think we can inline - the check you described is still done above

makramkd replied:
Oh but yeah good point about the default. Though the default would be `allowOutOfOrderExecution == false`

makramkd replied:
Ah I think you're saying that we should revert if OOO is enforced but not provided, which is the case when extraArgs is not provided at all. I think that's right but I wanna double check with Pieter to make sure this is the expected logic.

makramkd replied:
Yup this logic here is incorrect, to avoid a different default we should revert. Updating.

RensR commented on contracts/src/v0.8/ccip/test/e2e/End2End.t.sol:
```diff
@@ -93,14 +93,19 @@ contract E2E is EVM2EVMOnRampSetup, CommitStoreSetup, EVM2EVMOffRampSetup {
 
   function sendRequest(uint64 expectedSeqNum) public returns (Internal.EVM2EVMMessage memory) {
     Client.EVM2AnyMessage memory message = _generateTokenMessage();
+    message.extraArgs =
```
This change does nothing, right? This is already the default?

RensR commented on contracts/src/v0.8/ccip/test/offRamp/EVM2EVMOffRamp.t.sol:
```diff
@@ -1459,7 +1562,7 @@ contract EVM2EVMOffRamp__releaseOrMintTokens is EVM2EVMOffRampSetup {
   /// forge-config: ccip.fuzz.runs = 1024
   // Uint256 gives a good range of values to test, both inside and outside of the eth address space.
-  function test_Fuzz__releaseOrMintTokens_AnyRevertIsCaught_Success(uint256 destPool) public {
+  function test_Fuzz_releaseOrMintTokens_AnyRevertIsCaught_Success(uint256 destPool) public {
```
The double __ was intended here, the function to be tested is `_releaseOrMintTokens` and we always add an extra `_` before, so that makes 2

makramkd committed:
correctly enforce ooo if config says so

makramkd committed:
push validation logic to _validateMessage

makramkd committed:
forge fmt

makramkd committed:
Merge branch 'ccip-develop' into feat/ccip/ooo

makramkd committed:
Merge branch 'ccip-develop' into feat/ccip/ooo

makramkd committed:
bump optimizations back to 4k

looks like with solidity 8.24 we can bring back
the 4k optimizations that were set before these
OOO changes.

makramkd committed:
remove most fuzz tests

makramkd committed:
add tests specifically for OOO logic

makramkd committed:
add test cases to TestExecution_buildBatch

makramkd committed:
rename GetEVMExtraArgsV2 arg

makramkd committed:
forge fmt

RensR committed:
golf parsing and rm unused imports

makramkd committed:
regen wrappers

0xnogo commented on contracts/src/v0.8/ccip/offRamp/EVM2EVMOffRamp.sol:
```diff
@@ -357,12 +359,13 @@ contract EVM2EVMOffRamp is IAny2EVMOffRamp, AggregateRateLimiter, ITypeAndVersio
       // FAILURE   -> SUCCESS  no nonce bump
-      if (originalState == Internal.MessageExecutionState.UNTOUCHED) {
+      if (message.nonce > 0 && originalState == Internal.MessageExecutionState.UNTOUCHED) {
         s_senderNonce[message.sender]++;
```
It seems like we are not incrementing the `senderNonce` if ooo is enabled.
Are we using the Offramp's nonce other than to preserve the order of sequencing (i.e. accounting)?

makramkd replied:
The offramp nonce is always used to enforce ordering of ordered messages (i.e `message.nonce > 0`). For unordered messages we don't use the senderNonce because we don't care about enforcing any particular ordering on the execution of the message, just that its never replayed.

0xnogo commented on contracts/src/v0.8/ccip/onRamp/EVM2EVMOnRamp.sol:
```diff
@@ -370,14 +374,21 @@ contract EVM2EVMOnRamp is IEVM2AnyOnRamp, ILinkAvailable, AggregateRateLimiter,
+  function _fromBytes(bytes calldata extraArgs) internal view returns (Client.EVMExtraArgsV2 memory) {
     if (extraArgs.length == 0) {
-      return Client.EVMExtraArgsV1({gasLimit: i_defaultTxGasLimit});
+      return Client.EVMExtraArgsV2({gasLimit: i_defaultTxGasLimit, allowOutOfOrderExecution: false});
```
nit: `false` could also be an immutable variable. It might slightly improve consistency and readability.

RensR replied:
A variable that's always false? I'm not sure I see the value there

0xnogo commented on contracts/src/v0.8/ccip/test/StructFactory.sol:
```diff
@@ -176,7 +176,8 @@ contract StructFactory {
       defaultTokenDestGasOverhead: DEFAULT_TOKEN_DEST_GAS_OVERHEAD,
-      defaultTokenDestBytesOverhead: DEFAULT_TOKEN_BYTES_OVERHEAD
+      defaultTokenDestBytesOverhead: DEFAULT_TOKEN_BYTES_OVERHEAD,
+      enforceOutOfOrder: false
```
nit: same here, we could have it as const variable.

makramkd committed:
Merge branch 'ccip-develop' into feat/ccip/ooo
